File: exercicios/projeto-final/skeletons/rickandmorty/kmp/pratica/androidApp/src/main/kotlin/com/puciec/rickmortykmp/ListScreen.kt
package com.puciec.rickmortykmp

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.FilterChip
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import com.puciec.rickmortykmp.data.RickAndMortyApi
import com.puciec.rickmortykmp.data.CharacterSummary
import com.puciec.rickmortykmp.data.capitalize
import kotlin.coroutines.cancellation.CancellationException

private val STATUSES = listOf("alive", "dead", "unknown")

@Composable
fun ListScreen(api: RickAndMortyApi, onSelect: (Int) -> Unit) {
    var all by remember { mutableStateOf<List<CharacterSummary>>(emptyList()) }
    var loading by remember { mutableStateOf(true) }
    var error by remember { mutableStateOf<String?>(null) }
    var searchText by remember { mutableStateOf("") }
    var selectedStatus by remember { mutableStateOf<String?>(null) }
    var statusNames by remember { mutableStateOf<Set<String>?>(null) }
    
    LaunchedEffect(Unit) {
        try {
            all = api.fetchList()
        } catch (t: Throwable) {
            error = "Não foi possível carregar personagens."
        } finally {
            loading = false
        }
    }

    val filtered = all
        .filter { character -> statusNames?.contains(character.name) ?: true }
        .filter { character -> character.name.contains(searchText.trim(), ignoreCase = true) }

    if (loading) {
        Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) { CircularProgressIndicator() }
        return
    }
    if (error != null) {
        Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) { Text(error!!) }
        return
    }

    Box(Modifier.fillMaxSize().testTag("item-list-screen")) {
        LazyColumn(Modifier.fillMaxSize()) {
            item {
                OutlinedTextField(
                    value = searchText,
                    onValueChange = { searchText = it },
                    label = { Text("Buscar personagem") },
                    modifier = Modifier.fillMaxWidth().padding(12.dp).testTag("search-input"),
                )
            }
            item {
                LazyRow(
                    Modifier.fillMaxWidth().padding(horizontal = 12.dp),
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                ) {
                    item {
                        FilterChip(
                            selected = selectedStatus == null,
                            onClick = { selectedStatus = null; statusNames = null },
                            label = { Text("Todos") },
                            modifier = Modifier.testTag("category-chip-all"),
                        )
                    }
                    items(STATUSES) { status ->
                        FilterChip(
                            selected = selectedStatus == status,
                            onClick = { selectedStatus = status },
                            label = { Text(capitalize(status)) },
                            modifier = Modifier.testTag("category-chip-$status"),
                        )
                    }
                }
            }
            items(filtered, key = { it.id }) { character ->
                Text(
                    text = capitalize(character.name),
                    modifier = Modifier
                        .fillMaxWidth()
                        .testTag("item-card-${character.id}")
                        .clickable { onSelect(character.id) }
                        .padding(16.dp),
                    style = MaterialTheme.typography.bodyLarge,
                )
            }
        }
    }

    LaunchedEffect(selectedStatus) {
        statusNames = try {
            selectedStatus?.let { api.fetchNamesByStatus(it) }
        } catch (c: CancellationException) {
            throw c
        } catch (t: Throwable) {
            null
        }
    }
}

File: exercicios/projeto-final/skeletons/rickandmorty/flutter/pratica/lib/screens/list_screen.dart
import 'package:flutter/material.dart';
import '../api/rickandmorty_api.dart';
import '../models/character.dart';
import 'detail_screen.dart';

const List<String> _statuses = ['alive', 'dead', 'unknown'];

class ListScreen extends StatefulWidget {
  const ListScreen({super.key});

  @override
  State<ListScreen> createState() => _ListScreenState();
}

class _ListScreenState extends State<ListScreen> {
  final RickAndMortyApi _api = RickAndMortyApi();

  List<CharacterSummary> _all = [];
  bool _loading = true;
  String? _error;
  String _searchText = '';
  String? _selectedStatus;
  Set<String>? _statusNames;

  @override
  void initState() {
    super.initState();
    _loadList();
  }

  Future<void> _loadList() async {
    // TODO 1 (feature 1 — lista): chamar `_api.fetchList()`, guardar o
    // resultado em `_all` e marcar `_loading = false` via setState. Tratar
    // erro com `_error` (mensagem) igual ao padrão do catch abaixo.
    setState(() {
      _loading = false;
    });
  }

  Future<void> _selectStatus(String? status) async {
    // TODO 4 (feature 4 — categoria/filtro): quando `status` não é null,
    // chamar `_api.fetchNamesByStatus(status)` e guardar em `_statusNames`
    // (setState). Quando `status` é null, limpar `_statusNames`.
    setState(() {
      _selectedStatus = status;
    });
  }

  List<CharacterSummary> get _filtered {
    // TODO 3 (feature 3 — busca) + TODO 4 (feature 4 — categoria): filtrar
    // `_all` por `_statusNames` (quando não-nulo, manter só quem está no
    // set) e por `_searchText` (substring case-insensitive do `name`).
    return _all;
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Rick and Morty')),
      body: Semantics(
        identifier: 'item-list-screen',
        child: _buildBody(),
      ),
    );
  }

  Widget _buildBody() {
    if (_loading) {
      return const Center(child: CircularProgressIndicator());
    }
    if (_error != null) {
      return Center(child: Text(_error!));
    }
    return Column(
      children: [
        Padding(
          padding: const EdgeInsets.all(12),
          child: Semantics(
            identifier: 'search-input',
            child: TextField(
              decoration: const InputDecoration(
                labelText: 'Buscar personagem',
                border: OutlineInputBorder(),
                prefixIcon: Icon(Icons.search),
              ),
              onChanged: (value) => setState(() => _searchText = value),
            ),
          ),
        ),
        SizedBox(
          height: 44,
          child: ListView(
            scrollDirection: Axis.horizontal,
            padding: const EdgeInsets.symmetric(horizontal: 12),
            children: [
              _buildStatusChip(null, 'Todos'),
              for (final status in _statuses) _buildStatusChip(status, status),
            ],
          ),
        ),
        const SizedBox(height: 8),
        Expanded(
          child: ListView.builder(
            itemCount: _filtered.length,
            itemBuilder: (context, index) {
              final character = _filtered[index];
              return Semantics(
                identifier: 'item-card-${character.id}',
                child: ListTile(
                  title: Text(capitalize(character.name)),
                  onTap: () {
                    Navigator.of(context).push(
                      MaterialPageRoute(
                        builder: (_) => DetailScreen(characterId: character.id),
                      ),
                    );
                  },
                ),
              );
            },
          ),
        ),
      ],
    );
  }

  Widget _buildStatusChip(String? status, String label) {
    final selected = _selectedStatus == status;
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 4),
      child: Semantics(
        identifier: status == null ? 'category-chip-all' : 'category-chip-$status',
        child: ChoiceChip(
          label: Text(capitalize(label)),
          selected: selected,
          onSelected: (_) => _selectStatus(status),
        ),
      ),
    );
  }
}
